ocpcertificates: add ability to rotate MCS CA/cert

yuqi-zhang opened the pull request:
New oc command to rotate the Machine-Config-Server ca + cert. Also see: https://github.com/openshift/oc/pull/1449

To test this, take any cluster and run the command. Afterwards, try scaling a node into the cluster (e.g. on AWS, patch a worker machineset to add 1 more replica)

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -177,5 +187,78 @@ func run(o *RegenerateMCOOptions, args []string) error {
+
+	// Get user-data-secret
+	mapiSecrets := clientset.CoreV1().Secrets(mapiNamespace)
+	secretList, err := mapiSecrets.List(ctx, metav1.ListOptions{})
```
So I think this is fine, but we will be trying to edit the `user-data-managed` secrets generated/owned by the MCO today that are unused.  I think the MCO will edit them back?

In the general case as far as I understand it, probably what we should be doing is iterating over the *machineset* objects, and then we can find references to their user data from there.

yuqi-zhang replied:
Right, to rotate those, we'd have to change the MCO code I think. Let's check if we would need to do so or not

yuqi-zhang commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -177,5 +187,78 @@ func run(o *RegenerateMCOOptions, args []string) error {
+			}
+
+			// This is uh, quite unsavory, but arguably beats doing nested ignition version parsing?
+			if len(ign["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})) == 1 {
```
ack, will consider improving this, thanks!

cgwalters commented:
In my testing, I did have to bounce the MCS pods via `oc delete pods -l k8s-app=machine-config-server`.

cgwalters commented:
Another thing we could do here is make this more idempotent by adding an annotation to the secret with the last generation version of the input CA secret.  Right now we edit the userdata every run even when it'd be a no-op.

yuqi-zhang commented:
> In my testing, I did have to bounce the MCS pods via oc delete pods -l k8s-app=machine-config-server.

Ah hmm, I think I may have tested twice or something, I'll add that to the script

> Another thing we could do here is make this more idempotent by adding an annotation to the secret with the last generation version of the input CA secret. Right now we edit the userdata every run even when it'd be a no-op.

That is an option, although I doubt most people who would be using this script would run it that regularly (and really, it might be helpful to refresh the user-data regardless, just in case the user has made other manual modifications they would like to have the script auto-rectify)

cgwalters commented:
https://opensource.com/article/22/9/dynamically-update-tls-certificates-golang-server-no-downtime looks relevant; the default MCS logic here just loads the key once at startup. So we can and should change it to watch for rotation, but for now bouncing the pods seems OK.

cgwalters commented:
Ah even better, look at https://github.com/openshift/hypershift/blob/7cfb7f32a4b72c03a3a1e1ed552fe8b9a731dfee/ignition-server/cmd/start.go#L264 :arrow_right:  https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/certwatcher#CertWatcher

cgwalters commented:
@yuqi-zhang minor cleanup, how about de-indenting things here?

```
diff --git a/pkg/cli/admin/ocpcertificates/regeneratemco/command.go b/pkg/cli/admin/ocpcertificates/regeneratemco/command.go
index 5e20375cd..af28890b4 100644
--- a/pkg/cli/admin/ocpcertificates/regeneratemco/command.go
+++ b/pkg/cli/admin/ocpcertificates/regeneratemco/command.go
@@ -219,42 +219,42 @@ func run(o *RegenerateMCOOptions, args []string) error {
 		// 1. have a userData field
 		// 2. is an ignition config
 		userData, exists := secret.Data[userDataKey]
-		if exists {
-			// userData is an ignition config. To save the effort of multiple-version parsing, just parse it as a json
-			var userDataIgn interface{}
-			err = json.Unmarshal(userData, &userDataIgn)
-			if err != nil {
-				return fmt.Errorf("failed to unmarshal decoded user-data to json (secret %s): %w", secret.Name, err)
-			}
-			ignMap := userDataIgn.(map[string]interface{})
-			ign, isIgn := ignMap["ignition"].(map[string]interface{})
-			if !isIgn {
-				// Didn't find "ignition in user-data, warn but continue"
-				fmt.Printf("Unable to find ignition in user-data, skipping secret %s\n", secret.Name)
-				continue
-			}
-
-			// This is uh, quite unsavory, but arguably beats doing nested ignition version parsing?
-			if len(ign["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})) == 1 {
-				ca := userDataIgn.(map[string]interface{})["ignition"].(map[string]interface{})["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})[0]
-				ca.(map[string]interface{})["source"] = "data:text/plain;charset=utf-8;base64," + encodedCACert
-				fmt.Printf("Successfully updated for secret %s\n", secret.Name)
-			} else {
-				return fmt.Errorf("Additional CAs detected, cannot modify automatically. Aborting")
-			}
-
-			// Now marshal back to json
-			updatedIgnition, err := json.Marshal(userDataIgn)
-			if err != nil {
-				return fmt.Errorf("failed to marshal updated ignition back to json (secret %s): %w", secret.Name, err)
-			}
-
-			secret.Data[userDataKey] = updatedIgnition
-			_, err = mapiSecrets.Update(ctx, &secret, metav1.UpdateOptions{})
-			if err != nil {
-				return fmt.Errorf("could not update secret %s: %w", secret.Name, err)
-			}
+		if !exists {
+			continue
+		}
+		// userData is an ignition config. To save the effort of multiple-version parsing, just parse it as a json
+		var userDataIgn interface{}
+		err = json.Unmarshal(userData, &userDataIgn)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal decoded user-data to json (secret %s): %w", secret.Name, err)
+		}
+		ignMap := userDataIgn.(map[string]interface{})
+		ign, isIgn := ignMap["ignition"].(map[string]interface{})
+		if !isIgn {
+			// Didn't find "ignition in user-data, warn but continue"
+			fmt.Printf("Unable to find ignition in user-data, skipping secret %s\n", secret.Name)
+			continue
+		}
+
+		// This is uh, quite unsavory, but arguably beats doing nested ignition version parsing?
+		if len(ign["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})) == 1 {
+			ca := userDataIgn.(map[string]interface{})["ignition"].(map[string]interface{})["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})[0]
+			ca.(map[string]interface{})["source"] = "data:text/plain;charset=utf-8;base64," + encodedCACert
+			fmt.Printf("Successfully updated for secret %s\n", secret.Name)
+		} else {
+			return fmt.Errorf("Additional CAs detected, cannot modify automatically. Aborting")
+		}
+
+		// Now marshal back to json
+		updatedIgnition, err := json.Marshal(userDataIgn)
+		if err != nil {
+			return fmt.Errorf("failed to marshal updated ignition back to json (secret %s): %w", secret.Name, err)
+		}
 
+		secret.Data[userDataKey] = updatedIgnition
+		_, err = mapiSecrets.Update(ctx, &secret, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("could not update secret %s: %w", secret.Name, err)
 		}
 	}
 
```

yuqi-zhang commented:
Rebased + updated based on suggestions above. Removing draft for further review

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	openshiftOrg = "openshift"
+
+	// legacyRootCANamespace is the namespace of the legacy root CA
+	legacyRootCANamespace = "kube-system"
```
we placed this in kube-system before?  How naughty of us.

stlaz replied:
Unused. Go through all the consts and remove unused references.

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+
+// AddFlags registers flags for a cli
+func (o *RegenerateMCOOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&o.mcoNamespace, "mco-namespace", mcoNamespace, "Operate instead on this namespace for the MCO (useful for testing)")
```
hide this flag.

yuqi-zhang replied:
We can probably remove this flag, since the command is a bit more stable, and the testing should not otherwise break cluster functionality

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+}
+
+func NewCmdRegenerateTopLevel(restClientGetter genericclioptions.RESTClientGetter, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &RegenerateMCOOptions{
```
set the `mcoNamespace` here.

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	if !exists {
+		return fmt.Errorf("could not find MCS CA cert at %s after rotation", CASecretCertName)
+	}
+	encodedCACert := base64.StdEncoding.EncodeToString(caCert)
```
Get this from the configmap in the rotator above, not the secret itself.

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+		ign, isIgn := ignMap["ignition"].(map[string]interface{})
+		if !isIgn {
+			// Didn't find "ignition in user-data, warn but continue"
+			fmt.Printf("Unable to find ignition in user-data, skipping secret %s\n", secret.Name)
```
Fprintf to the iostream.  Doing that allows composition and unit testing for output if desired at some point.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+)
+
+const (
+	OneYear    = 365 * 24 * time.Hour
```
unexport

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+			continue
+		}
+
+		if len(ign["security"].(map[string]interface{})["tls"].(map[string]interface{})["certificateAuthorities"].([]interface{})) == 1 {
```
all the other constants and this one is where you draw the line? ;)

Can you use `unstructured.NestedString` to make this slightly more readable?

yuqi-zhang replied:
I had a comment about how unsavoury this is but then I removed it :D I'll make it a constant at least

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	// signerName is the name injected into the certificate as the signer
+	signerName = "oc-for-machine-config-server"
+	// openshiftOrg is the openshift organizational unit
+	openshiftOrg = "openshift"
```
unused

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	// mcsTlsSecretName is created by the installer and is not owned by default
+	mcsTlsSecretName = mcsName + "-tls"
+	// signerName is the name injected into the certificate as the signer
+	signerName = "oc-for-machine-config-server"
```
unused

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	}
+
+	// Get user-data-secret
+	mapiSecrets := clientset.CoreV1().Secrets(mapiNamespace)
```
Based on usage and potential failures, I think everything from here down is logically a different command.

cgwalters replied:
Hmm.  Yes, I see your point.  On the other hand:

- these secret objects are admin owned and not expected to change by anything else
- The cert code is already a "run once" controller and hence is idempotent and can be run multiple times reliably
- We really do want to update both of these things in "one go" as our default stance; doing one but not the other will result in a broken state

deads2k replied:
@yuqi-zhang I'm ok with that approach.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+type RegenerateMCOOptions struct {
+	RESTClientGetter genericclioptions.RESTClientGetter
+
+	mcoNamespace   string
```
why is this parametrized?

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+		Long:                  regenerateMCOLong,
+		Example:               regenerateMCOExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(run(o, args))
```
typically, `Run()` is a method of the options, that get `Validate()`-ed before use

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+		return err
+	}
+
+	host, err := oconfig.ConfigV1().Infrastructures().Get(context.TODO(), "cluster", metav1.GetOptions{})
```
get a real context here, should be passed to `run()`

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+	inf.Start(ch)
+	inf.WaitForCacheSync(ch)
+
+	ctx := context.WithValue(context.Background(), certrotation.RunOnceContextKey, true)
```
wire in a proper context from `run()`

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+		}
+		// userData is an ignition config. To save the effort of multiple-version parsing, just parse it as a json
+		var userDataIgn interface{}
+		err = json.Unmarshal(userData, &userDataIgn)
+		if err != nil {
```
nit: could be one line

stlaz reviewed (changes requested):
/lgtm cancel

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+			Client:        clientset.CoreV1(),
+			EventRecorder: recorder,
+		},
+		nil, // no operatorclient needed
```
No, that's what the `certrotation.RunOnceContextKey` does.  (We've tested this code)

stlaz replied:
I see, interesting

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+			Validity:  keyExpiry,
+			Refresh:   keyRefresh,
+			CertCreator: &certrotation.ServingRotation{
+				Hostnames: func() []string { return []string{apiserverIntURL.Hostname()} },
```
Yes

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -97,6 +114,8 @@ func NewCmdRegenerateTopLevel(restClientGetter genericclioptions.RESTClientGette
 func (o *RegenerateMCOOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().StringVar(&o.mcoNamespace, "mco-namespace", mcoNamespace, "Operate instead on this namespace for the MCO (useful for testing)")
+	cmd.Flags().BoolVar(&o.modifyUserData, "update-user-data", true, "If true, automatically update user-data secrets in machine-api namespace. Not useful if node scaling not backed by MachineSet.")
+	cmd.Flags().BoolVar(&o.outputNewCert, "output-new-cert", false, "If true, outputs the new CA cert for ignition use. Useful if need to manually update UPI clusters.")
```
Hmm, probably somewhere under

https://docs.openshift.com/container-platform/4.13/machine_management/index.html
perhaps
https://docs.openshift.com/container-platform/4.13/machine_management/user_infra/adding-compute-user-infra-general.html

But I don't think we need to block on this, we can just remove this code for now.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	fmt.Println("Successfully modified user-data secrets.")
```
this should print to `o.Out`, too

yuqi-zhang commented:
Ok, updated it to have two entrypoints, hopefully I've addressed all the comments

yuqi-zhang commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	cmd := &cobra.Command{
+		Use:                   "update-mco-user-data",
```
Any ideas for a better name here?

stlaz replied:
To get the answer to that question, I 'd need to answer this other one: what kind of data are we talking about here?

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+	ignFieldTLS      = "tls"
+	ignFieldCA       = "certificateAuthorities"
+	ignFieldSource   = "source"
+	ignDataPrefix    = "data:text/plain;charset=utf-8;base64,"
```
OK as is, but it would be cleaner to vendor github.com/vincent-petithory/dataurl  in the future I think as we do in the MCO and butane.

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+	cmd := &cobra.Command{
+		Use:                   "regenerate-mco",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Regenerate the machine config operator certificates in an OpenShift cluster"),
```
I wonder if we should clarify by s/operator/server/ here?  But eh.

yuqi-zhang replied:
I think for most users, the higher level "mco certificate" is hand-wavy enough to point them to the right path :)

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	cmd := &cobra.Command{
+		Use:                   "update-mco-user-data",
```
Minor bikeshed; maybe `update-machineset-user-data`?  Technically it's not fully owned by the MCO today.  Or maybe `update-ignition-pointer-config`?  

Eh, fine as is :smile:

yuqi-zhang replied:
Yeah, I mused about that a bit, I think it's better to keep the whole chain referencing the MCO, so it's easier for a non-in-depth user to understand the connection

But yeah, probably not the best naming

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,99 @@
+		updatedIgnition, err := json.Marshal(userDataIgn)
+		if err != nil {
+			return fmt.Errorf("failed to marshal updated ignition back to json (secret %s): %w", secret.Name, err)
+		}
```
Might be nice to do:
```
if bytes.Equal(userData, updatedIgnition) {
  fmt.Fprintf(o.IOStreams.Out, "Nothing to do for %s\n", secret.Name)
  continue
}
```
or so

cgwalters commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,99 @@
+		}
+	}
+
+	fmt.Fprintf(o.IOStreams.Out, "Successfully modified user-data secrets\n")
```
I think it'd be a bit clearer to move this print after we call `Update` to print each time we change a secret; would pair well with the above suggestion.

cgwalters reviewed (approved):
Overall LGTM, just super minor nits.  Will leave open for a bit in case anyone else has anything, if not then I'll drop lgtm later today or tomorrow.

yuqi-zhang commented:
Updated to match Colin's suggestions above

cgwalters commented:
OK I've looked through and it looks like you've adjusted for all the review comments so far.  Awesome!  I think we can do any other fixues as followups.
/lgtm

yuqi-zhang commented:
/retest-required

seems like flakes

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+	regenerateMCOLong = templates.LongDesc(`
+		Regenerate the Machine Config Operator certificates for an OCP v4 cluster.
+
+		More information about these certificates are in the product documentation;
+		visit [docs.openshift.com](https://docs.openshift.com)
+		and select the version of OpenShift you are using.
```
Remove this part please. From reading this, I do not know where to look, still. Reading the docs if something's unclear is what we'd expect from our users even if we don't mention it here.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/rotatecerts.go:
```diff
@@ -0,0 +1,123 @@
+		return fmt.Errorf("unable to get cluster infrastructure resource: %w", err)
+	}
+	if host.Status.APIServerInternalURL == "" {
+		return fmt.Errorf("no APIServerInternalURL found in cluster infrastructure resource")
```
much better now :+1:

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/rotatecerts.go:
```diff
@@ -0,0 +1,123 @@
+	)
+
+	ch := make(chan struct{})
+	inf.Start(ch)
+	inf.WaitForCacheSync(ch)
```
use `ctx.Done()`

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/rotatecerts.go:
```diff
@@ -0,0 +1,123 @@
+		return err
+	}
+
+	fmt.Fprintf(o.IOStreams.Out, "Successfully rotated MCS CA + certs\n")
```
Explain what's happening next so that the user is not confused as to why the command hasn't finished yet

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func (o *RegenerateMCOOptions) updateSecrets(ctx context.Context) error {
+	clientConfig, err := o.RESTClientGetter.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+	clientset, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
```
nit: if you had a `Complete()` method, you could store these in your `RegenerateMCOOptions` and would not have to reinitialize here in case of a call from `func (o *RegenerateMCOOptions) Run(ctx context.Context) error {`

It would also help you with unit testing where you'd be able to supply your fake clients to the method.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,105 @@
+
+	caCert, exists := mcsSecret.Data[corev1.TLSCertKey]
+	if !exists {
+		return fmt.Errorf("could not find MCS CA cert at %s after rotation", corev1.TLSCertKey)
```
```suggestion
		return fmt.Errorf("could not find MCS CA cert at %s", corev1.TLSCertKey)
```
Maybe the user did not perform the rotation yet.

Also, the `%s` is not helpful here, you'll want to call out the location of the secret, too.

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,105 @@
+			continue
+		}
+
+		caSlice, isSlice, err := unstructured.NestedFieldNoCopy(userDataIgn.(map[string]interface{}), ignFieldIgnition, ignFieldSecurity, ignFieldTLS, ignFieldCA)
```
```suggestion
		caSlice, isSlice, err := unstructured.NestedFieldNoCopy(userDataIgn.(map[string]interface{}), ignCAPath...)
```
where
```
ignCAPath = []string{ignFieldIgnition, "security", "tls",  "certificateAuthorities"}
```
defined above instead of the constants

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,105 @@
+			return fmt.Errorf("failed to find certificateauthorities field in ignition (secret %s): %w", secret.Name, err)
+		}
+		if len(caSlice.([]interface{})) == 1 {
+			caSlice.([]interface{})[0].(map[string]interface{})[ignFieldSource] = dataurl.EncodeBytes(caCert)
```
`unstructured.SetNestedStringSlice()` should work here I think

stlaz replied:
nvm, types are probably correct

yuqi-zhang replied:
Yes, sorry, so the data looks something like this:

```
"security":{"tls":{"certificateAuthorities":[{"source":"data:text/plain;charset=utf-8;base64,...","verification":{}}]}}
```

i.e. certificateAuthorities is a [] of ca objects as defined by ignition. Short of vendoring ignition, and doing multiple versions of parsing (ignition is very strict) this was the simplest way I could come up with. I opted not to use `SetNestedStringSlice` since it's not a string slice in here

stlaz commented on pkg/cli/admin/ocpcertificates/regeneratemco/updatesecrets.go:
```diff
@@ -0,0 +1,105 @@
+		if !isSlice || err != nil {
+			return fmt.Errorf("failed to find certificateauthorities field in ignition (secret %s): %w", secret.Name, err)
+		}
+		if len(caSlice.([]interface{})) == 1 {
```
nit: invert the condition to reduce nesting:
```
if len(caSlice.([]interface{})) > 1 {
    return fmt.Errorf("additional CAs detected, cannot modify automatically. Aborting")
}
<set the string slice here>
```

stlaz commented on vendor/github.com/vincent-petithory/dataurl/LICENSE:
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
```
please move vendoring into a separate commit

stlaz reviewed (changes requested):
/lgtm cancel

Could we also add unit tests, please? I did not realize that yesterday, sorry :neutral_face:

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+type RegenerateMCOOptions struct {
+	RESTClientGetter genericclioptions.RESTClientGetter
+
+	modifyUserData bool
```
followup: CLIs expose all provided fields for external composition

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/command.go:
```diff
@@ -0,0 +1,128 @@
+
+// AddFlags registers flags for a cli
+func (o *RegenerateMCOOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().BoolVar(&o.modifyUserData, "update-user-data", o.modifyUserData, "If true, automatically update user-data secrets in machine-api namespace. Not useful if node scaling not backed by MachineSet.")
```
```suggestion
	cmd.Flags().BoolVar(&o.modifyUserData, "update-ignition", o.modifyUserData, "If true, automatically update user-data secrets in machine-api namespace. Not useful if node scaling not backed by MachineSet.")
```

cgwalters replied:
It's both...it's ignition in user data.  The actual API objects we are modifying are called e.g. `worker-user-data`.

deads2k commented on pkg/cli/admin/ocpcertificates/regeneratemco/rotatecerts.go:
```diff
@@ -0,0 +1,123 @@
+	// TODO maybe add a watcher to make sure the MCS daemonset is ready here
+
+	if o.modifyUserData {
+		return o.updateSecrets(ctx)
```
```suggestion
		return o.RunUserDataUpdate(ctx)
```

?

deads2k commented:
/approve

yuqi-zhang commented:
Rebased, renamed the commands, updated to use ctx.Done, hopefully addressed the other issues, and re-tested to be still working on a fresh cluster.

Thanks for all the review!

cgwalters commented:
/lgtm
